pallets/asset-manager/src/lib.rs

```diff
@@ -0,0 +1,377 @@
+
+	/// Get the next available AssetId.
+	#[pallet::storage]
+	#[pallet::getter(fn next_asset_id)]
+	pub type NextAssetId<T: Config> = StorageValue<_, T::AssetId, ValueQuery>;
+
+	/// XCM transfer cost for different asset.
+	#[pallet::storage]
+	pub type UnitsPerSecond<T: Config> = StorageMap<_, Blake2_128Concat, T::AssetId, u128>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Register a new asset in the asset manager.
+		///
+		/// * `origin`: Caller of this extrinsic, the acess control is specfied by `ForceOrigin`.
+		/// * `location`: Location of the asset.
+		/// * `metadata`: Asset metadata.
+		/// * `min_balance`: Minimum balance to keep an account alive, used in conjunction with `is_sufficient`.
+		/// * `is_sufficient`: Whether this asset needs users to have an existential deposit to hold
+		///  this asset.
+		///
+		/// # <weight>
+		/// TODO: get actual weight
+		/// # </weight>
+		#[pallet::weight(50_000_000)]
+		#[transactional]
+		pub fn register_asset(
+			origin: OriginFor<T>,
+			location: T::AssetLocation,
+			metadata: T::AssetRegistrarMetadata,
+		) -> DispatchResult {
+			T::ModifierOrigin::ensure_origin(origin)?;
+			ensure!(
+				!LocationAssetId::<T>::contains_key(&location),
+				Error::<T>::LocationAlreadyExists
+			);
+			let asset_id = Self::get_next_asset_id()?;
+			T::AssetRegistrar::create_asset(
+				asset_id,
+				metadata.min_balance(),
+				metadata.clone().into(),
+				metadata.is_sufficient(),
+			)
+			.map_err(|_| Error::<T>::ErrorCreatingAsset)?;
+			AssetIdLocation::<T>::insert(&asset_id, &location);
+			AssetIdMetadata::<T>::insert(&asset_id, &metadata);
+			LocationAssetId::<T>::insert(&location, &asset_id);
+			Self::deposit_event(Event::<T>::AssetRegistered {
+				asset_id,
+				asset_address: location,
+				metadata,
+			});
+			Ok(())
+		}
+
+		/// Update an asset by its asset id in the asset manager.
+		///
+		/// * `origin`: Caller of this extrinsic, the acess control is specfied by `ForceOrigin`.
+		/// * `asset_id`: AssetId to be updated.
+		/// * `location`: `location` to update the asset location.
+		/// # <weight>
+		/// TODO: get actual weight
+		/// # </weight>
+		#[pallet::weight(50_000_000)]
+		#[transactional]
+		pub fn update_asset_location(
+			origin: OriginFor<T>,
+			#[pallet::compact] asset_id: T::AssetId,
+			location: T::AssetLocation,
+		) -> DispatchResult {
+			// checks validity
+			T::ModifierOrigin::ensure_origin(origin)?;
+			ensure!(
+				AssetIdLocation::<T>::contains_key(&asset_id),
+				Error::<T>::UpdateNonExistAsset
+			);
+			ensure!(
+				!LocationAssetId::<T>::contains_key(&location),
+				Error::<T>::LocationAlreadyExists
+			);
+			// change the ledger state.
+			let old_location =
+				AssetIdLocation::<T>::get(&asset_id).ok_or(Error::<T>::UpdateNonExistAsset)?;
+			LocationAssetId::<T>::remove(&old_location);
+			LocationAssetId::<T>::insert(&location, &asset_id);
+			AssetIdLocation::<T>::insert(&asset_id, &location);
```

Comment:
why insert new one instead of modifying it?
And if insert new one, you should remove previous one.